In [40]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
from matplotlib.mlab import PCA as mlabPCA
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA 

In [41]:
# Replace the path with the correct path for your data.
y2015 = pd.read_csv(
    'https://www.dropbox.com/s/0so14yudedjmm5m/LoanStats3d.csv?dl=1',
    skipinitialspace=True,
    header=1
)

C:\Users\09142640\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [42]:
# Convert ID and Interest Rate to numeric.
y2015['id'] = pd.to_numeric(y2015['id'], errors='coerce')
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

# Drop other columns with many unique variables
y2015.drop(['url', 'emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state', 'desc'], 1, inplace=True)

In [43]:
y2015 = y2015[:-2]

In [44]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()

In [45]:
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)


Using PCA below

In [46]:

sklearn_pca = PCA(n_components=10)
X_pca = sklearn_pca.fit_transform(X)


In [47]:
#using PCA for all the columns
cross_val_score(rfc, X_pca, Y, cv=10)

array([0.70423405, 0.77003633, 0.76913396, 0.77203106, 0.671147  ,
       0.64837331, 0.53280771, 0.53449381, 0.53475669, 0.571605  ])

In [48]:
#removing all the columns related to payment amount or outstanding principal
x = X.drop(['last_pymnt_amnt', 'out_prncp', 'last_pymnt_d_Jan-2017', 'out_prncp_inv'], 1)
cross_val_score(rfc, x, Y, cv=10)

array([0.98026644, 0.98000522, 0.97641946, 0.97986274, 0.97121824,
       0.97416291, 0.84532998, 0.97653708, 0.97560976, 0.9685793 ])

After removing all the columns related to payment amount or outstanding principal, we have around 98% accuracy 